### DISCUSSION TITLE
Passing a custom ts config to ts node

### DISCUSSION THREAD

**Discussion by olup** [pulumi/pulumi#9053] [Pulumi in Practice] [answered]
I have a problem as I want to use ts 4.4 for my code but pulumi uses 3.7. some options throws when starting pulumi that I need in my code (built with esbuild). I'd like to keep the main tsconfig for the code and have a separate config for pulumi. Is there a param to give a specific tsconfig file to pulumi's ts-node ? Or a workaround for this, update pulumi typescript, define its ts-node, etc... ?
Thanks !

**Comment by rawkode** [comment]
Pulumi uses whatever version of `ts-node` is defined in `package.JSON` (AFAIK). 

However, you can use different `tsconfig.JSON` files. 

Defining `TS_NODE_PROJECT=path/to/tsconfig` should encourage Pulumi to use an alternative. 

You can also modify the `Pulumi.yaml` with args to node:

https://github.com/pulumi/pulumi/blob/master/tests/integration/nodejs/esm-ts/Pulumi.yaml

**Reply by olup** [reply]
I am doing `TS_NODE_PROJECT=tsconfig.pulumi.json pulumi up -y -f`but pulumi is still using `tsconfig.json`. I'll try the pulumi.yaml way but i didn't understand very well how it's used by ts-node. Plus none of this is documented anywhere ?

**Reply by olup** [reply]
And from the stack trace of the error, pulumi is using its own `ts-node` :
````
    error: TSError: ⨯ Unable to compile TypeScript:
    error TS5023: Unknown compiler option 'useUnknownInCatchVariables'.
    
        at createTSError (/Users/olup/dev/pr/pr-serverless/node_modules/@pulumi/pulumi/node_modules/ts-node/src/index.ts:261:12)
        at Object.register (/Users/olup/dev/pr/pr-serverless/node_modules/@pulumi/pulumi/node_modules/ts-node/src/index.ts:265:42)
        at Object.run (/Users/olup/dev/pr/pr-serverless/node_modules/@pulumi/cmd/run/run.ts:176:16)
        at /Users/olup/dev/pr/pr-serverless/node_modules/@pulumi/cmd/run/index.ts:160:57
```

**Reply by olup** [reply]
According to the code and a lot of trial and error, it seems passing both `PULUMI_NODEJS_TSCONFIG_PATH=./tsconfig.pulumi.json TS_NODE_PROJECT=./tsconfig.pulumi.json ` seems to work...

**Reply by Frassle** [reply]
Oh interesting that you had to set both, did you try just setting `PULUMI_NODEJS_TSCONFIG_PATH` by itself?

**Reply by Frassle** [reply]
Oh actually see https://github.com/pulumi/pulumi/pull/9062 I think tsnode wasn't picking up that custom named project file and we weren't passing it in. I think that PR should fix this so you just need to set `PULUMI_NODEJS_TSCONFIG_PATH`.

**Reply by olup** [reply]
Yep, exactly what I came to as well. When should this be merged ?

**Reply by Frassle** [reply]
Assuming our build issues get fixed soon (see https://github.com/goreleaser/goreleaser/issues/2943) I'd hope we can get this merged in today.

**Reply by nimbinatus** [reply]
I believe this PR has been merged. Has it fixed the problem for you, @olup?

**Reply by olup** [reply]
Thanks ! Let me check that. On Holliday now but will be back in two weeks, and might try before. Thanks for this swift PR

**Comment by nimbinatus** [comment, marked as answer]
I'm going to mark this one as answered based on the thread and the merged https://github.com/pulumi/pulumi/pull/9062. Please let us know if this should get reopened!
